File: typescript/examples/planner-streaming.ts
/** Deterministic fixture only: real Adapters must forward actual Provider records. */
import {
  Agent, InMemoryAgentAdapters, ModelWorkPlanner,
  type ModelGateway, type ModelCapabilitySnapshot, type OutputEvent,
} from "purra";

const capabilities: ModelCapabilitySnapshot = {
  schemaVersion: 1, profileId: "example:planner", providerProtocol: "custom",
  contextWindowTokens: 32768, maxCallOutputTokens: 512, thinkingTokenAccounting: "unknown",
  protocol: { reasoningControl: "selectable", reasoningReplay: "ignored", toolCalling: "supported",
    requiredToolChoice: "supported", parallelToolCalls: "supported", streaming: "supported", cancellation: "supported",
    assistantContentWithToolCalls: "optional", jsonSchemaLevel: "unknown", streamFinishSemantics: "normalized", usageSemantics: "normalized" },
};
function gate() {
  let resolve!: () => void;
  const promise = new Promise<void>((done) => { resolve = done; });
  return { promise, resolve };
}
function require(value: unknown): asserts value { if (!value) throw new Error("Example assertion failed"); }
let release = gate();
const model: ModelGateway = {
  capabilities,
  async invoke() { throw new Error("This example uses managed streams"); },
  stream(request, signal) {
    const planning = request.messages.some((message) => message.attributes?.planningContract);
    return { ...(request.outputLimit === undefined ? {} : { appliedOutputLimit: request.outputLimit.maxTokens }),
      async *[Symbol.asyncIterator]() {
        if (planning) {
          yield { contentDelta: JSON.stringify({ v: 1, type: "progress", text: "I will check the request's scope." }) + "\n" };
          let cancel!: () => void;
          try {
            const stopped = new Promise<void>((resolve) => { cancel = resolve; signal?.addEventListener("abort", cancel, { once: true }); });
            await Promise.race([release.promise, stopped]);
            if (signal?.aborted) return;
          } finally { signal?.removeEventListener("abort", cancel); }
          yield { contentDelta: JSON.stringify({ v: 1, type: "plan", plan: { workPlan: {
            title: "Answer", steps: [{ id: "answer", title: "Answer", type: "review", executor: "model" }],
          } } }) + "\n" };
        } else yield { contentDelta: "The answer is ready." };
        yield { finishReason: "stop" as const };
      },
    };
  },
};

const storage = new InMemoryAgentAdapters(); // Replace runs for durable canonical storage.
const agent = new Agent({ model, runRepository: storage.runs, outputPublisher: storage.outputs,
  planning: { plannerFactory: (tasks) => new ModelWorkPlanner(tasks) },
});
const input = { messages: [{ role: "user" as const, content: "Give a concise answer." }], planningMode: "planned" as const };
const options = { budgets: { maxRunOutputTokens: null } };
const handle = await agent.submit(input, options);
const live: OutputEvent[] = [];
for await (const event of handle.events()) {
  live.push(event);
  if (event.kind === "planning.progress") {
    require((await storage.runs.listEvents(handle.runId, 0)).some((stored) => stored.eventId === event.eventId));
    release.resolve();
  }
}
require((await handle.result).output === "The answer is ready.");
const replay: OutputEvent[] = [];
for await (const event of handle.events()) replay.push(event);
require(JSON.stringify(live) === JSON.stringify(replay));
require(live.some((event) => event.kind === "plan.updated")); // Admitted summary; private plan remains in storage.

release = gate();
const canceled = await agent.submit(input, options);
const result = canceled.result.catch(() => undefined);
for await (const event of canceled.events()) {
  if (event.kind === "planning.progress") { await canceled.cancel(); break; }
}
await result;
require((await canceled.snapshot()).status === "canceled");
const before = await storage.runs.listEvents(canceled.runId, 0);
release.resolve();
require(JSON.stringify(before) === JSON.stringify(await storage.runs.listEvents(canceled.runId, 0)));

File: examples/python/planner_streaming.py
"""Deterministic public-API example: managed planning, subscription, cancel, replay.

The fixture Gateway replaces an external Provider; no network or paid calls.
Production Gateways must stream actual Provider records, never invent progress.
"""
from __future__ import annotations

import asyncio
import json
from dataclasses import replace

from purra.api import AgentCore, AgentPlanner, InMemoryAgentAdapters
from purra.contracts import (
    AgentMessage, AgentRunRequest, DomainContext, ModelRequest, ModelStream,
    ModelStreamChunk, PlanningConstraints, PlanningMode, RuntimeLimits,
)
from purra.model_protocol import generic_capability_snapshot


class Policy:
    def planning_constraints(self, request, capabilities):
        return PlanningConstraints(allow_model_only_fallback=False)


class FixtureGateway:
    def __init__(self):
        self.release_plan = asyncio.Event()

    async def complete(self, *args, **kwargs):
        raise AssertionError("This example uses managed streams")

    async def stream(self, messages, invocation, signal=None):
        planning = any("planning component" in message.content for message in messages)

        async def chunks():
            if planning:
                yield ModelStreamChunk(content_delta=json.dumps({
                    "v": 1, "type": "progress", "text": "I will check the request's scope."
                }) + "\n")
                await self.release_plan.wait()
                yield ModelStreamChunk(content_delta=json.dumps({
                    "v": 1, "type": "plan", "plan": {
                        "needsTodos": True, "title": "Answer", "todos": [
                            {"id": "answer", "title": "Answer", "type": "review", "executor": "model"}
                        ]
                    }
                }) + "\n")
            else:
                yield ModelStreamChunk(content_delta="The answer is ready.")
            yield ModelStreamChunk(finish_reason="stop")

        return ModelStream(chunks=chunks(), model="fixture",
                           applied_generation_limit=invocation.output_budget.max_generation_tokens)


async def exercise() -> None:
    storage = InMemoryAgentAdapters()  # Replace BOTH repositories for durable storage.
    gateway = FixtureGateway()
    core = AgentCore(model_gateway=gateway, planner=AgentPlanner(gateway),
                     planning_policy=Policy(), run_repository=storage.runs,
                     output_repository=storage.outputs, output_publisher=storage.publisher,
                     runtime_limits=RuntimeLimits(max_run_generation_tokens=None))
    request = AgentRunRequest(
        messages=(AgentMessage(role="user", content="Give a concise answer."),),
        model=ModelRequest(provider="fixture", model="fixture",
            capability_snapshot=replace(generic_capability_snapshot(), profile_id="example:planner",
                                        max_generation_tokens=512)),
        domain_context=DomainContext(namespace="example"), context_window=32_768,
        planning_mode=PlanningMode.PLANNED,
    )
    try:
        handle = await core.submit(request)
        live = []
        async for event in handle.subscribe():  # Public canonical stream by default.
            live.append(event)
            if event.kind.value == "planning.progress":
                persisted = await storage.outputs.list_events(handle.run_id, after_sequence=0)
                assert event in persisted  # Durable receipt before notification.
                assert not gateway.release_plan.is_set()
                gateway.release_plan.set()
        assert (await handle.wait()).status.value == "done"
        assert live == [event async for event in handle.subscribe()]  # Same IDs/sequence on replay.
        snapshots = await storage.runs.get(handle.run_id)
        assert snapshots.steps[0].id == "answer"  # Complete, validated plan only.

        gateway.release_plan.clear()
        canceled = await core.submit(request)
        async for event in canceled.subscribe():
            if event.kind.value == "planning.progress":
                await canceled.cancel("example cancellation")
                break
        assert (await canceled.wait()).status.value == "canceled"
        history = [event async for event in canceled.subscribe()]
        gateway.release_plan.set()
        assert history == [event async for event in canceled.subscribe()]
    finally:
        await core.close()


if __name__ == "__main__":
    asyncio.run(exercise())
